### PR TITLE
Migrate MusicBrainz plugin to MetaBrainz.MusicBrainz

### PR DESCRIPTION
**Pull request by MrTimscampi**
**Changes**

* Removes the custom XML serializer
* Use the MetaBrainz-recommended [MetaBrainz.MusicBrainz](https://github.com/Zastai/MetaBrainz.MusicBrainz/) library.
  This makes it a lot easier to get and do stuff with MusicBrainz data, removes a bunch of code on our end, handles rate limiting for us, has the MetaBrainz-approved user agent format and a bunch of other stuff.

**Notes**

***Important***: I wanted to preserve the ability to use a custom server, but the library works a bit differently. Users will need to change `https://musicbrainz.com` to `musicbrainz.com`. We should also maybe add stuff like "Enable SSL for requests" to set HTTP/HTTPS and a port setting, but I'm not quite sure what the best way to do this would be. Suggestions are welcome.

~~This is still work in progress, hence the draft.~~ This is now at feature parity with the current provider.

Through working on this, it became apparent that, as of now, the MusicBrainzAlbumProvider does essentially nothing aside from setting a few IDs. I do want to improve this, but I first want to get the plugin into a more workable state with the MetaBrainz library, as it's a pain to work with otherwise.

Release matching is currently a bit trash. Essentially, if it has a release ID, it uses that, which is great. But if it doesn't, it just takes the first release in the release group, or the first result of the search. That's incredibly naive and will lead to mismatched albums (which you can't see currently, since it doesn't use any of the info aside from the release date and annotation).  
I'd like to take a look at how Picard and Beets handle release matching, to improve the way we do it. Obviously, this will come after the migration is done.

I think there's some discussion to be had in regards to merging this plugin with the Cover Art Archive one, at some point in the future.  
Not only do we already get Cover Art Archive info in the API response we get here, but it could also improve the search experience for users by providing preview images and reducing the complexity of having to use multiple plugins.  
I would hazard a guess that this would be done only after moving the MusicBrainz provider out of the server repository, though.

**Issues**

### CODE REVIEW

File: MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzArtistProvider.cs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        private Query _musicBrainzQuery;
+
+        public MusicBrainzArtistProvider()
+        {
+            _musicBrainzQuery = new Query();
+
+            Current = this;
+        }
+
+        internal static MusicBrainzArtistProvider Current { get; private set; }
```
Comment (crobibero) [inline MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzArtistProvider.cs:37]: Where is this used?

Comment (MrTimscampi) [reply MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzArtistProvider.cs:37]: Not sure, it was there before, so I just assumed it was from the album provider and was there before (I just cleaned up the artist provider to have all the standard fields like name and such).

I assumed it was needed by the plugin engine or something.

If it's not, we can remove it from both providers.

Comment (cvium) [reply MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzArtistProvider.cs:37]: It's just a manual way of handling singleton classes. If it has no references you can remove it afaik.

File: MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzArtistProvider.cs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                    return GetResultsFromResponse(artistSearchResults.Results.Select(result => result.Item).ToList());
+                }
+
+                if (searchInfo.Name.HasDiacritics())
+                {
+                    // Try again using the search with an accented characters query
+                    var artistAccentsSearchResults = await _musicBrainzQuery.FindArtistsAsync($"artistaccent:\"{searchInfo.Name}\"");
+                    if (artistAccentsSearchResults.Results.Count > 0)
+                    {
+                        return GetResultsFromResponse(artistAccentsSearchResults.Results.Select(result => result.Item).ToList());
```
Comment (cvium) [inline MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzArtistProvider.cs:69]: I'm generally not a fan of the use of `GetResultsFromResponse`. You should have a method that takes a single `IArtist` and returns a new `RemoteSearchResult` and `GetResultsFromResponse` should yield return this method.

In the current design there's some wasted allocations. You're calling ToList (list is allocated) and then calling Select (enumerator is created and then further up the call stack ToList is called again), which is incorrect use of IEnumerable. And for single values you're allocating a new array, which is a waste.

**Comment by MrTimscampi** [comment]
Marking as blocked while we wait for the (seemingly upcoming) license change from MS-PL to MIT for the library.

**Comment by CodeMason** [comment]
This library is now MIT Licensed - any update when this issue can be re-evaluated?

**Comment by Shadowghost** [comment]
#7603 is the continuation of this PR.

**Pull request closed without merging**